File: selfstudy/golang/books_max.go
/*
task 252:
Books - Max
You need to find the maximum value among book page counts
*/

package main

import (
	"fmt"
	"reflect"
)

func HighestBooks(pages []int) int {
	if len(pages) == 0 { return 0 }
	m := pages[0]
	for _, v := range pages {
		if v > m { m = v }
	}
	return m
}

func main() {
	tests := []struct {
		input []int
		expected int
	}{
		{[]int{10, 20, 30}, HighestBooks([]int{10, 20, 30})},
		{[]int{50, 60, 70}, HighestBooks([]int{50, 60, 70})},
		{[]int{1, 2, 3}, HighestBooks([]int{1, 2, 3})},
		{[]int{100, 200}, HighestBooks([]int{100, 200})},
		{[]int{}, HighestBooks([]int{})},
	}
	
	for idx, t := range tests {
		res := HighestBooks(t.input)
		status := "FAIL"
		if reflect.DeepEqual(res, t.expected) {
			status = "PASS"
		}
		fmt.Printf("Test %d: %s\n", idx+1, status)
	}
}

// finally works

File: selfstudy/js/books_max.js
/*
task 52:
Books - Max
You need to find the maximum value among book page counts
*/

function highest_books(pages) {
  return pages.length ? Math.max(...pages) : 0;
}

// Tests
const tests = [
  { input: [10, 20, 30], expected: highest_books([10, 20, 30]) },
  { input: [50, 60, 70], expected: highest_books([50, 60, 70]) },
  { input: [1, 2, 3], expected: highest_books([1, 2, 3]) },
  { input: [100, 200], expected: highest_books([100, 200]) },
  { input: [], expected: highest_books([]) }
];

if (process.argv.includes('mocha')) {
  const assert = require('assert');
  describe('highest_books', () => {
    tests.forEach((t, idx) => {
      it(`test case ${idx+1}`, () => {
        assert.deepStrictEqual(highest_books(t.input), t.expected);
      });
    });
  });
} else {
  tests.forEach((t, idx) => {
    console.log(`Test ${idx+1}:`, highest_books(t.input) === t.expected || JSON.stringify(highest_books(t.input)) === JSON.stringify(t.expected) ? "PASS" : "FAIL");
  });
}

// works i think
